terminate/terminate.go

```diff
@@ -0,0 +1,209 @@
+	}
+
+	return sampledSectors
+}
+
+func TerminateSectors(ctx context.Context, api *lotusapi.FullNodeStruct, minerAddr address.Address, sectors *bitfield.BitField, ts *types.TipSet) (*TerminateSectorResult, error) {
+	// create a new terminate sector result
+	result := TerminateSectorResult{
+		InitialPledgeFromSample:      big.NewInt(0),
+		TerminationFeeFromSample:     big.NewInt(0),
+		AvgInitialPledgePerSector:    big.NewInt(0),
+		AvgTerminationFeePerPledge:   big.NewInt(0),
+		EstimatedTotalInitialPledge:  big.NewInt(0),
+		EstimatedTotalTerminationFee: big.NewInt(0),
+		InitialPledge:                big.NewInt(0),
+		SectorCount:                  0,
+	}
+
+	_, mstate, err := mstat.LoadMinerActor(ctx, api, minerAddr, ts)
+	if err != nil {
+		return nil, err
+	}
+
+	minerInfo, err := mstate.Info()
+	if err != nil {
+		return nil, err
+	}
+
+	smoothedPow, err := util.TotalPowerSmoothed(ctx, api, ts)
+	if err != nil {
+		return nil, err
+	}
+
+	smoothedRew, err := util.ThisEpochRewardsSmoothed(ctx, api, ts)
+	if err != nil {
+		return nil, err
+	}
+
+	sectorInfos, err := api.StateMinerSectors(context.Background(), minerAddr, sectors, ts.Key())
+	if err != nil {
+		return nil, err
+	}
+
+	// if no sectors found return empty result
+	if len(sectorInfos) == 0 {
+		return &result, nil
+	}
+
+	for _, s := range sectorInfos {
+		sectorPower := miner8.QAPowerForSector(minerInfo.SectorSize, util.ConvertSectorType(s))
+
+		// the termination penalty calculation
+		termFee := miner8.PledgePenaltyForTermination(
+			s.ExpectedDayReward,
+			ts.Height()-s.PowerBaseEpoch,
+			s.ExpectedStoragePledge,
+			smoothedPow,
+			sectorPower,
+			smoothedRew,
+			s.ReplacedDayReward,
+			s.PowerBaseEpoch-s.Activation,
+		).Int
+
+		// this should never happen, but termination fee should never be negative
+		if termFee.Cmp(big.NewInt(0)) < 0 {
+			return nil, xerrors.Errorf("negative termination fee: %v", termFee)
+		}
+
+		if s.InitialPledge.Int.Cmp(big.NewInt(0)) == 0 {
+			// if the initial pledge is 0, something is up?
+			return nil, xerrors.Errorf("0 initial pledge for sector")
+		}
+
+		result.TerminationFeeFromSample = big.NewInt(0).Add(result.TerminationFeeFromSample, termFee)
+		result.InitialPledgeFromSample = big.NewInt(0).Add(result.InitialPledgeFromSample, s.InitialPledge.Int)
+		result.SectorCount++
+	}
+
+	// calculate the average penalty per pledge as result.TerminationFee / result.InitialPledge and keep the reult in big.int
+	result.AvgTerminationFeePerPledge = util.DivWad(result.TerminationFeeFromSample, result.InitialPledgeFromSample)
+	// log average penalty per pledge
+	// log.Printf("AVERAGE PENALTY PER PLEDGE: %f\n", util.ToFIL(result.AvgTerminationFeePerPledge))
+
+	// calculate the average initial pledge per sector as result.InitialPledge / result.SectorCount and keep the result in big.int
+	result.AvgInitialPledgePerSector = big.NewInt(0).Div(result.InitialPledgeFromSample, big.NewInt(int64(result.SectorCount)))
```

Comment:
Should use `SampledSectorCount` after the type change

Reply:
done